csrc/multidevice/utils.cpp

```diff
@@ -151,6 +222,108 @@ void insertReshardings(Fusion* fusion) {
   }
 }
 
+void insertShardedAxisReordering(Fusion* fusion) {
+  auto exprs = fusion->exprs();
+  std::vector<Expr*> reshard_exprs;
+  for (auto expr : exprs) {
+    if (isResharding(expr)) {
+      reshard_exprs.push_back(expr);
+    }
+  }
+  for (auto expr : reshard_exprs) {
+    NVF_ERROR(
+        ir_utils::isTvOp(expr),
+        "Non-tv op is not supported yet: ",
+        expr->toString());
+    NVF_ERROR(
+        expr->outputs().size() == 1,
+        "Resharding operations can only have one output");
+    NVF_ERROR(
+        expr->inputs().size() == 1,
+        "Resharding operations can have only one input");
+    auto output = expr->outputs().at(0)->as<TensorView>();
+    auto input = expr->inputs().at(0)->as<TensorView>();
+    auto [shard_additions, shard_deletions] = allocationShardings(expr);
+    NVF_ERROR(
+        shard_additions.size() + shard_deletions.size() <= 1,
+        "Resharding expr can only support one axis")
+
+    // For gather operations i.e. ID goes from sharded to unsharded
+    // this will rematerialize a sharded axis.
+    // ProcessGroup expects contiguous tensors.
+    // Update input to push the rematerialized axis to the front -> collective
+    // -> permute the rematerizlied axis to the proper location Example: [i0
+    // DIDx(i1)] -> [i0 i1] Rewritten to: [DIDx(i1) i0] -> [i1 i0] -> [i0 i1]
+    // Note: there are no reduction based collectives that materializes an axis
+    // so expr is guaranteed to be a set.
+    if (!shard_deletions.empty()) {
+      auto id = shard_deletions[0];
+      int idx = static_cast<int>(input->domain()->posOf(id));
+      if (isContiguousShard(input, id)) {
+        continue;
+      }
+      TensorView* input_permute = permute(input, {{idx, 0}});
+      TensorView* output_permute = set(input_permute);
+      TensorView* new_output = permute(output_permute, {{0, idx}});
+      ir_utils::replaceValInAllExprInputsAndFusionOutputs(output, new_output);
+      for (int i = 0; i < static_cast<int>(output->nDims()); i++) {
+        new_output->axis(i)->parallelize(output->axis(i)->getParallelType());
+      }
+      new_output->setDeviceMesh(output->getDeviceMesh());
+      shardAllLike(input, {input_permute});
+      shardAllLike(new_output, {output_permute});
+      fusion->removeExpr(expr);
+    }
+    // For scatter operations i.e. ID goes from unsharded to sharded
+    // Update input to push the scattered axis to the front -> collective ->
+    // permute the sharded axis to the proper location.
+    // Example: [i0 i1 DIDx(i2)] -> [i0 DIDx(i1) r(DIDx(i2))]
+    // Rewritten to: [i0 i1 DIDx(i2)] -> [i1 i0 DIDx(i2)] ->
+    //                    [DIDx(i1) i0 r(DIDx(i2))] -> [i0 DIDx(i1)]
```

Comment:
This is a bit confusing as we do not use the same parallel type for multiple domains of the same tensor. 

Is this a reduce-scatter pattern?

Reply:
Can you update the example to follow the nvFuser convention, i.e., don't use the same parallel type for multiple domains in the same tensor.

Reply:
Yes, I can do that. In fact the code assumes that the reduced ID is not device parallel. So lowering to reduce scatter is looking for patterns of this form:
[I0 DIDx(i1) i2] -> [DIDx(i0) r(i1) i2]

We had some earlier discussions about how to represent reduction of a Device parallel iter domain and whether it should be `DIDx(i1) -> r(i1)` vs. `DIDx(i1) -> r(DIDx(i1))`. 
There were some cases where the second one was preferred. Since we pattern match onto a collective, we don't control how that reduction is implemented so these two examples are lowered onto the same collective. 
It sounds like we should use the first option to avoid two axes with DIDx in reduce scatter scenarios?

Reply:
The latter is the correct parallel reduction pattern in nvFuser. I don't think we can express the reduce-scatter pattern as a single nvFuser op. What we could do instead is to have a parallel reduction op followed by another parallel scatter op. Then when they were lowered, we could lower them to a single reduce-scatter op.

Generally speaking, we prefer to have only primitive ops in the Fusion IR, so for example, we don't have a "all-reduce" op, but it's expressed as a reduction followed by a broadcast. This pattern can be lowered to an all-reduce Kernel IR op when it's possible. See for example, https://github.com/NVIDIA/Fuser/blob/main/csrc/kernel_ir.h#L1004.

Reply:
Makes sense, thanks for the explanation. 
I agree as primitives, reduce scatter would be expressed as parallel reduction op + parallel scatter op. 

For now, I'll assume it's ok that we have `set` with one to one lowering to "fused" MPI collectives. Later we will probably update as we improve the communication lowering analysis.